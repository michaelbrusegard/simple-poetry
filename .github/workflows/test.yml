name: Test

on:
  push:
    branches:
      - main

jobs:
  test:
    name: Run Tests
    runs-on: self-hosted
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '22'

      - name: Install dependencies
        run: npm install

      - name: Run tests
        run: npx vitest run
