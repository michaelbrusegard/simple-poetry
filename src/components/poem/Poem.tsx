import styles from './Poem.module.css'
import { useNavigate } from 'react-router-dom'
import { useEffect, useState } from 'react'
import { useSuspenseQuery } from '@tanstack/react-query'
import { Button } from '../../components/ui/Button'
import { PoemProps } from '../../types'

function Poem({
  poemTitle,
  poemAuthor,
}: {
  poemTitle: string
  poemAuthor: string
}) {
  const navigate = useNavigate()
  const poem = useSuspenseQuery({
    queryKey: ['poem', poemTitle, poemAuthor],
    queryFn: async () => {
      const response = await fetch(
        `https://poetrydb.org/author,title/${poemAuthor};${poemTitle}/lines,author,linecount`,
      )
      return response.json()
    },
  })

  const [isFavorite, setIsFavorite] = useState(false)

  useEffect(() => {
    const favorites = JSON.parse(localStorage.getItem('favorites') || '[]')
    setIsFavorite(
      favorites.some(
        (fav: PoemProps) =>
          fav.title === poemTitle &&
          fav.author === poem.data[0].author &&
          fav.linecount === poem.data[0].linecount,
      ),
    )
  }, [poem])

  const handleFavoriteClick = () => {
    const favorites = JSON.parse(localStorage.getItem('favorites') || '[]')
    if (isFavorite) {
      const newFavorites = favorites.filter(
        (fav: PoemProps) =>
          !(
            fav.title === poemTitle &&
            fav.author === poem.data[0].author &&
            fav.linecount === poem.data[0].linecount
          ),
      )
      localStorage.setItem('favorites', JSON.stringify(newFavorites))
    } else {
      favorites.push({
        title: poemTitle,
        author: poem.data[0].author,
        linecount: poem.data[0].linecount,
      })
      localStorage.setItem('favorites', JSON.stringify(favorites))
    }
    setIsFavorite(!isFavorite)
  }
  return (
    <>
      <h2 className={styles.author}>{poem.data[0].author}</h2>
      <p className={styles.lines}>{poem.data[0].lines.join('\n')}</p>
      <div className={styles.buttonWrapper}>
        <Button variant='secondary' onClick={() => navigate(-1)}>
          Back
        </Button>
        <Button
          variant={isFavorite ? 'destructive' : 'default'}
          onClick={handleFavoriteClick}
        >
          {isFavorite ? 'Remove Favorite' : 'Favorite'}
        </Button>
      </div>
    </>
  )
}

export { Poem }
