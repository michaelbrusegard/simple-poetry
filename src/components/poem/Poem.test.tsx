import { waitFor, fireEvent, act } from '@testing-library/react'
import { render } from '../../test/utils'
import { Poem } from './Poem'

describe('Poem', () => {
  it('renders without crashing', () => {
    render(
      <Poem
        poemTitle='Spring and Winter ii'
        poemAuthor='William Shakespeare'
      />,
    )
  })

  it('fetches and displays poem data', async () => {
    const { getByText } = render(
      <Poem
        poemTitle='Spring and Winter ii'
        poemAuthor='William Shakespeare'
      />,
    )

    await waitFor(() => {
      expect(getByText(/WHEN icicles hang by the wall/i)).toBeInTheDocument()
      expect(getByText(/When all aloud the wind doe blow/i)).toBeInTheDocument()
    })
  })

  it('handles favorite button click', async () => {
    const { getByText } = render(
      <Poem
        poemTitle='Spring and Winter ii'
        poemAuthor='William Shakespeare'
      />,
    )

    const favoriteButton = getByText(/Favorite/i)
    fireEvent.click(favoriteButton)

    await waitFor(() => {
      expect(getByText(/Remove Favorite/i)).toBeInTheDocument()
    })

    fireEvent.click(favoriteButton)

    await waitFor(() => {
      expect(getByText(/Favorite/i)).toBeInTheDocument()
    })
  })

  it('should match the snapshot', async () => {
    let fragment
    await act(async () => {
      const { asFragment } = render(
        <Poem
          poemTitle='Spring and Winter ii'
          poemAuthor='William Shakespeare'
        />,
      )
      fragment = asFragment()
    })
    expect(fragment).toMatchSnapshot()
  })
})
