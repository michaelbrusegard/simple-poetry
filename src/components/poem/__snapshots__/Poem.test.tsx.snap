// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Poem > should match the snapshot 1`] = `
<DocumentFragment>
  <p
    class="_lines_4abd61"
  >
    WHEN icicles hang by the wall,
   And Dick the shepherd blows his nail,
And Tom bears logs into the hall,
   And milk comes frozen home in pail,
When blood is nipp'd, and ways be foul,
Then nightly sings the staring owl,
   To-whit!
To-who!--a merry note,
While greasy Joan doth keel the pot.

When all aloud the wind doe blow,
   And coughing drowns the parson's saw,
And birds sit brooding in the snow,
   And Marian's nose looks red and raw,
When roasted crabs hiss in the bowl,
Then nightly sings the staring owl,
   To-whit!
To-who!--a merry note,
While greasy Joan doth keel the pot.
  </p>
  <div
    class="_buttonWrapper_4abd61"
  >
    <button
      class="_button_8ea813 _variantSecondary_8ea813 _sizeDefault_8ea813 "
    >
      Back
    </button>
    <button
      class="_button_8ea813 _variantDefault_8ea813 _sizeDefault_8ea813 "
    >
      Favorite
    </button>
  </div>
</DocumentFragment>
`;
